### PR TITLE
Mark enum.Flag.name as Optional[str]

### LINKED ISSUE #5610: enum.Flag.name can be None
If you run the following program:
```py
from enum import Flag, auto
from typing import TYPE_CHECKING

class MyFlag(Flag):
    A = auto()
    B = auto()

both = MyFlag.A | MyFlag.B
print(f"{both} has name {both.name}")
if TYPE_CHECKING:
    reveal_type(both.name)
```
It outputs:
```
MyFlag.B|A has name None
```
However, if it is checked with mypy, the output reads:
```
testcase.py:11:17: note: Revealed type is "builtins.str"
```
I think that `enum.Flag` should have `name` overridden to be `Optional[str]` instead of just `str`.

Since `Flag` inherits its `name` attribute from `Enum`, strictly speaking either `Enum.name` should also have type `Optional[str]` or we have to accept that `Flag` does not comply with the interface of its parent class. I think the latter is the more pragmatic solution.

### PR DESCRIPTION
**Pull request by mthuurne**
Flags can be composed and not all composed values are guaranteed to have a name.

Closes #5610

### CODE REVIEW

File: stdlib/enum.pyi
Comment (Akuli) [inline stdlib/enum.pyi:61]: Maybe `str` instead of `int`?

```
>>> enum.Flag('Foo', ['A', 'B', 'C']).A.name
'A'
```

And a small nit: We prefer the new `Foo | None` syntax instead of `Optional[Foo], so you can do this:

```suggestion
    name: str | None  # type: ignore
```

**Review by Akuli — CHANGES_REQUESTED** [review]

File: stdlib/enum.pyi
Comment (mthuurne) [reply stdlib/enum.pyi]: Yes, obviously the name should be a string, thanks for spotting that. I also updated the syntax.

**Comment by mthuurne** [comment]
Something weird is happening in mypy after this change. If I add this to the test case:
```py
def use_flag(flag: MyFlag) -> None:
    reveal_type(flag.name)
    reveal_type(flag.value)
```
mypy will output, as expected:
```
testcase.py:15:17: note: Revealed type is "Union[builtins.str, None]"
testcase.py:16:17: note: Revealed type is "builtins.int"
```
However, if I change it to this instead:
```py
def use_flag(flag: Flag) -> None:
    reveal_type(flag.name)
    reveal_type(flag.value)
```
mypy will output:
```
testcase.py:15:17: note: Revealed type is "Union[builtins.str, None]"
testcase.py:16:17: note: Revealed type is "Union[builtins.str, None]"
```
Any idea what is going on here?

**Comment by Akuli** [comment]
Yes, it's a problem with the mypy plugin. I added some prints to it, and the problem is where it says `underlying_type = _first(proper_types)`. The way it works seems to be a bit of a hack, since it obviously should skip our `str | None`.

**Comment by Akuli** [comment]
This change fixes the mypy plugin:

```diff
diff --git a/mypy/plugins/enums.py b/mypy/plugins/enums.py
index ade32998e..5d40472ad 100644
--- a/mypy/plugins/enums.py
+++ b/mypy/plugins/enums.py
@@ -158,7 +158,9 @@ def enum_value_callback(ctx: 'mypy.plugin.AttributeContext') -> Type:
             if _implements_new(info):
                 return ctx.default_attr_type
 
-            stnodes = (info.get(name) for name in info.names)
+            # Members of ENUM_PREFIXES classes are not really enum members
+            names = [] if info.fullname in ENUM_PREFIXES else info.names
+            stnodes = (info.get(name) for name in names)
 
             # Enums _can_ have methods and instance attributes.
             # Omit methods and attributes created by assigning to self.*
```

I'm not going to pull request this to mypy, for several reasons:
- It takes forever to run mypy's tests, especially when there's several tests that need to be changed to make them pass.
- I don't know how to properly test this.
- The previous PR I submitted to mypy has soon been open for a year, with no replys from mypy maintainers.

But I'm fine with someone else creating a PR to mypy with this change. I won't complain that this is "my code" or whatever. We can't merge this in typeshed until the mypy plugin is fixed, because that would break mypy.

**Comment by mthuurne** [comment]
> But I'm fine with someone else creating a PR to mypy with this change. I won't complain that this is "my code" or whatever. We can't merge this in typeshed until the mypy plugin is fixed, because that would break mypy.

I've had better luck getting changes accepted in mypy, although it wasn't easy to get the tests to work they wanted to. I'll give it a go.

I think the patch still needs a bit of work though, as mypy now outputs:
```
testcase.py:15:17: note: Revealed type is "Union[builtins.str, None]"
testcase.py:16:17: note: Revealed type is "Any"
```
So while the output is no longer incorrect, it has become less precise.

I'll try to rework your patch and if successful, submit it.

**Comment by mthuurne** [comment]
While debugging, I found a solution that doesn't require changing mypy: if we annotate `enum.Flag.value`, the existing mypy enum plugin will use that annotated value.

**Comment by Akuli** [comment]
It seems to be fine to require the value of a flag to be `int`, as the docs say it should be "powers of two (1, 2, 4, 8, …)".

That said, I'm a bit worried about whether adding members to `Flag` breaks something else in the plugin, if we don't add code to handle it. I tried a few things and I couldn't find anything that breaks, but I'm not fully confident and it would be nice if someone else reviewed this too. Maybe @hauntsaninja?

**Comment by mthuurne** [comment]
> It seems to be fine to require the value of a flag to be `int`, as the docs say it should be "powers of two (1, 2, 4, 8, …)".

I checked the implementation of `enum.Flag` and it will only operate on `int` or a type that mimics it very closely, including being comparable to `0` and having a `bit_length()` method. So in practice, I think requiring `int` is fine.

> That said, I'm a bit worried about whether adding members to `Flag` breaks something else in the plugin, if we don't add code to handle it. I tried a few things and I couldn't find anything that breaks, but I'm not fully confident and it would be nice if someone else reviewed this too. Maybe @hauntsaninja?

There is an annotation for `value` in `enum.Enum` and `enum.IntEnum` already, so I think it's relatively safe. But it certainly wouldn't hurt to have another person look at it.

**Review by JelleZijlstra — APPROVED** [review]
Looks good to me, @Akuli do you have any further feedback?

**Review by Akuli — APPROVED** [review]

**Pull request merged**
